### PR TITLE
 Support on_entry and on_exit for states

### PR DESCRIPTION
**Pull request by MartinBroers**
Issue #60 describes what I needed. I may have sprinkled a few `Debug` statements here and there which may not have been required, but I am open to suggestions to remove them. 

I am opening this MR since @pleepleus asked to have a go at it. I don't want him to start from scratch.

### CODE REVIEW

File: macros/src/codegen.rs [hunk 1]
```diff
@@ -356,6 +370,10 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
                             .zip(out_states.iter().zip(guard_action_parameters.iter().zip(guard_action_ref_parameters.iter()))),
                     )
                     .map(|(guard, (action, (out_state, (g_a_param, g_a_ref_param))))| {
+                        let out_state_string = &out_state.to_string()[0..out_state.to_string().find('(').unwrap_or_else(|| out_state.to_string().len())];
```
Comment (ryan-summers) [inline macros/src/codegen.rs:373]: I think a [`split`](https://doc.rust-lang.org/std/primitive.str.html#method.split) is more appropriate and canonical here

File: macros/src/codegen.rs [hunk 2]
```diff
@@ -467,6 +493,10 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
             #guard_error
             #guard_list
             #action_list
+            #entry_list
+
+            #[allow(missing_docs)]
+            fn transition_callback(&self, new_state: &Option<#states_type_name>) {}
```
Comment (ryan-summers) [inline macros/src/codegen.rs:499]: Why a transition callback when we already have an on_entry and on_exit for all states? Is this basically just intended to be a default function?

File: macros/src/codegen.rs [hunk 3]
```diff
@@ -390,11 +412,15 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
...
                                 let _data = self.context.#a(#temporary_context_call #g_a_param) #action_await ;
+                                    self.context_mut().#exit_ident();
+                                    self.context_mut().#entry_ident();
                                 self.state = Some(#states_type_name::#out_state);
                             }
                         } else {
                             quote! {
                                 self.state = Some(#states_type_name::#out_state);
+                                    self.context_mut().#exit_ident();
+                                    self.context_mut().#entry_ident();
```
Comment (ryan-summers) [inline macros/src/codegen.rs:423]: I don't want to force everyone to implement these if they're not needed. Can we guard these calls based on their existence in the FSM definition?

**Review by ryan-summers — COMMENTED** [review]
Some general feedback is below. I don't know the state of this or if the intent is to merge it as-is or if it's just inteded to be a starting point.

If this is just a reference implementation for someone else to pick up, can we mark it as a draft? Otherwise, I'm not sure if I should be reviewing or not.

File: macros/src/codegen.rs (hunk 3 above)
Comment (MartinBroers) [reply macros/src/codegen.rs:423]: Yes, I have seen other functions to be behind these flags, will update this accordingly (and will put this MR in draft while I haven't done that).

When I have this flag and the other comments fixed, I think it is good to go.

File: macros/src/codegen.rs (hunk 2 above)
Comment (MartinBroers) [reply macros/src/codegen.rs:499]: Yes, understandable. Idea was a generic callback to capture the transition, I use it for bookkeeping in other parts of my application, synchronizing states to my backend. Could your on_entry_[stateA,stateB] and on_exit[stateA,stateB] for this as well, but I'd be duplicating the code for all transitions. So I thought a generic one is better.
Will hide this behind a feature flag as well.

**Comment by MartinBroers** [comment]
> Some general feedback is below. I don't know the state of this or if the intent is to merge it as-is or if it's just inteded to be a starting point.
> 
> If this is just a reference implementation for someone else to pick up, can we mark it as a draft? Otherwise, I'm not sure if I should be reviewing or not.

Yes, wasn't totally clear on that. Will work on this to make it final and ping you when I'm happy. Also, open for edits by others if they have time before I do.

File: macros/src/codegen.rs (hunk 1 above)
Comment (MartinBroers) [reply macros/src/codegen.rs]: Fixed. I think it is now more readable than I had it first, thanks.

**Comment by MartinBroers** [comment]
I have added two new features, `generate_entry_exit_states` and `generate_transition_callback`. Code will not be generated when these flags are not present. Please let me know what you think of these additions. I am a bit unsure on the naming of the whole thing, but naming is important.

File: macros/src/codegen.rs [hunk 4]
```diff
@@ -19,6 +19,9 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
     let state_machine_context_type_name =
         format_ident!("{sm_name}StateMachineContext", span = sm_name_span);
 
+    let generate_entry_exit_states = sm.generate_entry_exit_states;
```
[the remark this answers is not in the document]
Comment (ryan-summers) [reply macros/src/codegen.rs:22]: Then, if the `enter_*` and `exit_*` functions don't exist in the UML, we don't generate the respective calls.

[the remark this answers is not in the document]
Comment (ryan-summers) [reply macros/src/codegen.rs:22]: Or the notation you specified at the end of https://github.com/korken89/smlang-rs/issues/60#issue-1536434502 would work as well. If we're following Boost SML, that would be preferable.

**Comment by dgehriger** [comment]
@ryan-summers : do you know what the state of this PR is? Has it gone stale?

**Comment by ryan-summers** [comment]
I haven't seen any response from the author to my review, so I would assume it's been abandoned.

**Comment by MartinBroers** [comment]
@dgehriger I had it on my list to fix, but got other priorities to work on first. I spent some time to rework the comments, but I failed in properly understanding the macros. Then I got other things on my plate and this got pushed down on my todo list.

So, to re-open the issue for myself and start thinking about it again, I have the token parser to parse `<` and `>`, but I am failing to get those functions into the generated result. Its been a while since I've looked at it though.

So, in `transition.rs` I now have
```
if input.parse::<Token![<]>().is_ok() 
```
which generates an `entry` for `StateTransitions`. Now, I need to get those fields back in the result, if they are not None, then generate a function for them. My code is very much in progress, but I'll push it anyway, and I am very open to remarks about them.

**Comment by andygikling** [comment]
Bump.  This appears to be one of the more feature-rich state machines in rust that anyone uses.  on_enter and on_exit is table stakes for a state machine library.  Can we get this feature in here please?

**Comment by ryan-summers** [comment]
I have nothing against the implementation @andygikling.

This, like many open-sourced projects, is maintained primarily through volunteer time. I'm happy to review any proposed changes you may have for the repo, but I don't have significant amounts of time to implement new features unless I need them for specific projects that I'm working on.

**Comment by andygikling** [comment]
Ok thanks I'll take a closer look and see what's missing!

**Comment by MartinBroers** [comment]
So, I've finally gotten around to update this PR. I implemented the notation with `>` and `<` to indicate entry- and exit functions. 

@ryan-summers pinging for review.

**Comment by ryan-summers** [comment]
Would it make more sense to provide generic `on_enter_<snake-case statename>` and `on_exit_<snake-case statename>` for all states and provide them in the trait definition? Then, if the user wants to implement the entry/exit functions, they just have to provide a custom implementation.

This keeps the notation entirely out of the DSL and the compiler can optimize out the empty enter/exit calls if the user doesn't actually need them.

**Comment by MartinBroers** [comment]
@ryan-summers Thanks for the feedback. That was my original thought as well. When I read your comment I tried the flag and in rebasing the flag didn't work. So I updated it now and now both ways should work. I kinda like your suggestion with the '<' and '>' as well, so I left them in for now. Updated README and added an example.

**Comment by ryan-summers** [comment]
> @ryan-summers Thanks for the feedback. That was my original thought as well. When I read your comment I tried the flag and in rebasing the flag didn't work. So I updated it now and now both ways should work.

I was actually advocating for removing the flag entirely and having the `on_entry` and `on_exit` functions defined by-default for all users. The trait would provide the default implementation, so for example:
```rs
enum States {
    Start,
}

trait StateMachineContext {
    fn on_entry_start() {
        // Default, empty implementation that the user can override in their context definition.
    }
    fn on_exit_start() {
    }


}

**Comment by MartinBroers** [comment]
I have removed the flag and let all `on_entry` and `on_exit` functions be autogenerated from now. I think it is pretty neat now.

Thanks for your time. I will have a look at the rebase error and the pipeline failures.

**Review by ryan-summers — APPROVED** [review]

**Comment by ryan-summers** [comment]
@MartinBroers I removed the `transition_callback` function for now because I wanted to separate it from the `on_entry` and `on_exit` change. I'm not opposed to the idea, but I think it needs more refinement/discussion first, and the other callbacks are ready for now. Feel free to open up a secondary PR for the transition callback changes if you'd still like them!

I think I'd go for the route where it's defined as empty-by-default and the user doesn't need to specify any flags etc.

**Pull request merged**